### ISSUE TITLE
X11: No FL_RELEASE when releasing the forward/back mouse button

### ISSUE THREAD

**Issue by CendioHalim** [fltk/fltk#1076]
**Describe the bug**
I don't get FL_RELEASE when I click on the forward/back side buttons on my mouse. These buttons map to `xevent.xbutton.button` 8 and 9.

**To Reproduce**
1. Click on the forward or back mouse button.

**Expected behavior**
I should get an `FL_RELEASE` when I release the mouse button.

**FLTK Version**
This commit https://github.com/fltk/fltk/commit/2252ba98390505f353ea35e8ade3e3914b46c977 broke this behaviour.

A simple fix would be to explicitly check if the button is between 4-7:

```diff
diff --git a/src/Fl_x.cxx b/src/Fl_x.cxx
index dcae946ab..bd2843db2 100644
--- a/src/Fl_x.cxx
+++ b/src/Fl_x.cxx
@@ -2089,7 +2089,7 @@ int fl_handle(const XEvent& thisevent)
     Fl::e_keysym = FL_Button + xevent.xbutton.button;
     set_event_xy(window);
     Fl::e_state &= ~(FL_BUTTON1 << (xevent.xbutton.button-1));
-    if (xevent.xbutton.button > Button3) { // "buttons" 4-7 = mousewheel events: don't send FL_RELEASE
+    if (xevent.xbutton.button > 3 && xevent.xbutton.button < 8 ) { // "buttons" 4-7 = mousewheel events: don't send FL_RELEASE
       return 0;
     }
```

**Comment by CendioHalim** [comment]
Something else seems to be broken as well. If I hold down LMB and keep it pressed, then press and release either side button, I get an `FL_RELEASE` the patch above. But as soon as I move the mouse again, I will keep receiving `FL_PUSH` events for the side button until LMB is released.

**Comment by Albrecht-S** [comment]
@CendioHalim Thank you for the report and the proposed patch. Your simple patch is not going to work for several reasons, but it's a good start. However, we would need to extend some constants (enums) for `Fl::event_state()` like for instance `FL_BUTTONS`, but very likely even more. All mouse event related code would need to be checked as well.

Note that I could not find any X11 docs (and headers) that "know" of more than 5 mouse buttons, where buttons 4 and 5 represent the scroll wheel (forward and backward). Even "buttons" 6 and 7 that are definitely sent by current X11 software for "horizontal scroll wheel" events don't seem to be officially documented. I would appreciate if you or anybody else could give me a link to valid X11 docs on how these pseudo buttons are to be handled.

Reference: Xorg docs on [XButtonEvent](https://www.x.org/releases/X11R7.7/doc/man/man3/XButtonEvent.3.xhtml).
Citation: _"For the XButtonPressedEvent and XButtonReleasedEvent structures, this member is called button. It represents the pointer button that changed state and can be the Button1, Button2, Button3, Button4, or Button5 value."_

Currently the code in `case ButtonPress` does **_not expect_** any mouse button value greater than 7, but this is not tested in this code. If your observation that forward and backward mouse buttons send button values 8 and 9 is true, then the code in [case ButtonPress](https://github.com/fltk/fltk/blob/84eeac5892c47c657deb961ef3aa88076c8ad651/src/Fl_x.cxx#L2034) sets currently **undefined bits** in `Fl::event_state()` outside the range of `FL_BUTTONS` (Button1 .. Button3) which will never be considered internally for button states etc.. This may be the/one reason why your second comment "Something else seems to be broken..." can be observed.

That said, after a quick check of the code I imagine basically two solutions:

1. Bug fix only: for consistency and to prevent setting "undefined" bits in `Fl::event_state()` we need to filter button values `> 7` in the `ButtonPress` event (see link above) and keep the rest of the code as-is. This is not a solution for the reported issue but it fixes an inconsistency and avoids further bugs. Forward and backward mouse buttons (any button value `> 7`) would be ignored for now.
2. Check X11 docs (if available), check FLTK internals, extend enums appropriately, document the new features, and test, test, test ...

**_From my current point of view_** solution (1)  would be the preferred one for FLTK 1.4.0 since 1.4 is currently on feature freeze and planned to be released soon.

Solution (2) would be too big a change and too much error prone for a quick fix in FLTK 1.4.0 and it would change the user interface (API) - at least we'd have to add new public constants (extend enums as described above). This should be reserved for FLTK 1.5.0 or (if possible) a later 1.4.x release after thorough and successful tests. We'd also need to implement and test this on all other supported platforms (Windows, macOS, Wayland) to be sure we don't break anything.

FYI: I just ordered a new mouse with forward/backward buttons for future testing. ;-)

**Comment by Albrecht-S** [comment]
FTR: Although this issue is currently for X11 only (according to the title and the proposed patch), the real solution would require a cross platform implementation on all platforms. Therefore I didn't set **any** platform specific labels.

**Comment by Albrecht-S** [comment]
See also PR #1068 for a macOS and Windows implementation.

**Comment by CendioHalim** [comment]
> Note that I could not find any X11 docs (and headers) that "know" of more than 5 mouse buttons, where buttons 4 and 5 represent the scroll wheel (forward and backward). Even "buttons" 6 and 7 that are definitely sent by current X11 software for "horizontal scroll wheel" events don't seem to be officially documented. I would appreciate if you or anybody else could give me a link to valid X11 docs on how these pseudo buttons are to be handled.
> 
> Reference: Xorg docs on [XButtonEvent](https://www.x.org/releases/X11R7.7/doc/man/man3/XButtonEvent.3.xhtml). Citation: _"For the XButtonPressedEvent and XButtonReleasedEvent structures, this member is called button. It represents the pointer button that changed state and can be the Button1, Button2, Button3, Button4, or Button5 value."_

Indeed, it seems like X11 only keeps the state of the 5 mouse buttons as you say.

> Currently the code in `case ButtonPress` does **_not expect_** any mouse button value greater than 7, but this is not tested in this code. If your observation that forward and backward mouse buttons send button values 8 and 9 is true, then the code in [case ButtonPress](https://github.com/fltk/fltk/blob/84eeac5892c47c657deb961ef3aa88076c8ad651/src/Fl_x.cxx#L2034) sets currently **undefined bits** in `Fl::event_state()` outside the range of `FL_BUTTONS` (Button1 .. Button3) which will never be considered internally for button states etc.. This may be the/one reason why your second comment "Something else seems to be broken..." can be observed.

You're correct. If the goal is that the state for buttons other than Button1/2/3 should be kept correctly, this is not enough. Looking at how X11 works, it does not keep track of the states of "additional" buttons. If we want to mimic this behaviour, would it be enough to simply react to the press/release events? Or do we want to keep the state of them internally as well? If we  don't care about tracking the state, then simply doing what we do now (`Fl::e_keysym = FL_Button + xevent.xbutton.button;`), and filtering buttons > 7 would for `Fl::e_state` be enough.

> FYI: I just ordered a new mouse with forward/backward buttons for future testing. ;-)

Great! :)

**Comment by Albrecht-S** [comment]
FTR: commit c8dadeffe0661596e26d0bdd0ab899db86acb0e0 of branch `mouse-buttons` in [my fork](https://github.com/Albrecht-S/fltk/tree/mouse-buttons) should work with X11 and Wayland as well, i.e. this branch includes now support of `BACK` and `FORWARD` mouse buttons for all FLTK platforms.

@CendioHalim Please test if this fixes all issues for you. TIA.

**Comment by Albrecht-S** [comment]
@CendioHalim What is the state of this issue after PR #1081 has been merged?

Please note that I can't follow the full discussion and all potential changes, but I think that the literal issue _"No FL_RELEASE when releasing the forward/back mouse button"_ has been fixed with the merge mentioned above.

Can we close this issue as "fixed"?

**Comment by CendioHalim** [comment]
Indeed, this is fixed as of #1081.

**Comment by Albrecht-S** [comment]
@CendioHalim Thanks for confirmation.
